node-uniffi/src/lib.rs

```diff
@@ -0,0 +1,303 @@
+//! Native library providing Rust to mobile language bindings for the Lumina node.
+//!
+//! This crate uses Mozillas UniFFI to generate Swift and Kotlin bindings for the Lumina node,
+//! allowing it to be used from iOS and Android applications.
+#![cfg(not(target_arch = "wasm32"))]
+
+mod error;
+mod types;
+
+use celestia_types::ExtendedHeader;
+use error::{LuminaError, Result};
+use lumina_node::{
+    blockstore::RedbBlockstore,
+    events::{EventSubscriber, TryRecvError},
+    network::Network,
+    node::PeerTrackerInfo,
+    store::RedbStore,
+    Node,
+};
+use std::{str::FromStr, sync::Arc};
+use tendermint::hash::Hash;
+use tokio::sync::Mutex;
+use types::{NetworkInfo, NodeEvent, NodeStartConfig, PeerId, SyncingInfo};
+use uniffi::Object;
+
+uniffi::setup_scaffolding!();
+
+lumina_node::uniffi_reexport_scaffolding!();
+
+/// The main Lumina node that manages the connection to the Celestia network.
+#[derive(Object)]
+pub struct LuminaNode {
+    node: Arc<Mutex<Option<Node<RedbBlockstore, RedbStore>>>>,
+    network: Network,
+    events_subscriber: Arc<Mutex<Option<EventSubscriber>>>,
+}
+
+#[uniffi::export(async_runtime = "tokio")]
+impl LuminaNode {
+    /// Sets a new connection to the Lumina node for the specified network.
+    #[uniffi::constructor]
+    pub fn new(network: Network) -> Result<Arc<Self>> {
+        Ok(Arc::new(Self {
+            node: Arc::new(Mutex::new(None)),
+            network,
+            events_subscriber: Arc::new(Mutex::new(None)),
+        }))
+    }
+
+    /// Start the node without optional configuration.
+    /// UniFFI needs explicit handling for optional parameters to generate correct bindings for different languages.
+    pub async fn start(&self) -> Result<bool> {
+        self.start_with_config(None).await
+    }
```

Comment:
I think this should be remove and rename `start_with_config` to `start`. In general we should give control to the user when they have to choose non-optional configs. For example a user should explicitly set the `network`.

Reply:
I see. I thought that since the user already has to set the network in the `new` function, we can make a `start() `function that doesn't change anything in the rest of the optional configuration, but I'll change that. Apart from that, are the configuration options sufficient?

Reply:
Another approach, which I think we prefer, is to have the whole `NodeStartConfig` in `new` and then just have `start` (but not `start_with_config`). Also in this case it would make sense to rename the config struct to `NodeConfig`.

> Apart from that, are the configuration options sufficient?

Yes looks ok.

Reply:
> to have the whole NodeStartConfig in new and then just have start without the config

I'm curious. If we could also remove `start` and start the node right away within the `new`, and then make `stop` take `self`, we could get rid of `Mutex<Option<...>>` entirely.

I know we use the separate `new` and `start` pattern but that's because we use a `WebWorker` there. Would this work in case of uniffi? or the semantics of other languages like swift and kotlin don't play well with `stop(self)`?

Reply:
I think the main problem here is that uniffi automatically wraps objects in an Arc if there are functions that need ownership. We'd probably need a construction like 
```rust
pub async fn stop(self: Arc<Self>) ... {
   match Arc::try_unwrap(self) { 
      Ok(node) =>  node.node.stop().await
      ...
} 
```
But even then, I'm not sure if it would work. Should I give it a try?

Reply:
i agree, i played around with it, it's getting messy 😅